src/multirust-cli/self_update.rs

```diff
@@ -0,0 +1,969 @@
+        return Ok(());
+    }
+
+    new_path.push_str(";");
+    new_path.push_str(&old_path);
+    try!(environment.set_value("PATH", &new_path)
+         .map_err(|_| Error::PermissionDenied));
+
+    // Tell other processes to update their environment
+    unsafe {
+        SendMessageTimeoutA(HWND_BROADCAST,
+                            WM_SETTINGCHANGE,
+                            0 as WPARAM,
+                            "Environment\0".as_ptr() as LPARAM,
+                            SMTO_ABORTIFHUNG,
+                            5000,
+                            ptr::null_mut());
+    }
+
+    Ok(())
+}
+
+/// Decide which rcfiles we're going to update, so we
+/// can tell the user before they confirm.
+fn get_remove_path_methods() -> Result<Vec<PathUpdateMethod>> {
+    if cfg!(windows) {
+        return Ok(vec![PathUpdateMethod::Windows]);
+    }
+
+    let bashrc = env::home_dir().map(|p| p.join(".bashrc"));
+    let zshrc = env::home_dir().map(|p| p.join(".zshrc"));
+    let kshrc = env::home_dir().map(|p| p.join(".kshrc"));
+
+    let rcfiles = vec![bashrc, zshrc, kshrc];
+    let existing_rcfiles = rcfiles.into_iter()
+        .filter_map(|f|f)
+        .filter(|f| f.exists());
+
+    let export_str = try!(shell_export_string());
+    let matching_rcfiles = existing_rcfiles
+        .filter(|f| {
+            let file = utils::read_file("rcfile", f).unwrap_or(String::new());
+            let ref addition = format!("\n{}", export_str);
+            file.contains(addition)
+        });
+
+    Ok(matching_rcfiles.map(|f| PathUpdateMethod::RcFile(f)).collect())
+}
+
+#[cfg(windows)]
+fn do_remove_from_path(methods: &[PathUpdateMethod]) -> Result<()> {
+    assert!(methods.len() == 1 && methods[0] == PathUpdateMethod::Windows);
+
+    use winreg::RegKey;
+    use winapi::*;
+    use user32::*;
+    use std::ptr;
+
+    let root = RegKey::predef(HKEY_CURRENT_USER);
+    let environment = try!(root.open_subkey_with_flags("Environment", KEY_READ | KEY_WRITE)
+                           .map_err(|_| Error::PermissionDenied));
+
+    let old_path: String = environment.get_value("PATH").expect("non-unicode PATH");
+    let ref path_str = try!(get_cargo_home()).join("bin").to_string_lossy().to_string();
+    let idx = if let Some(i) = old_path.find(path_str) {
+        i
+    } else {
+        return Ok(());
+    };
+
+    let mut new_path = old_path[..idx].to_string();
+    new_path.push_str(&old_path[idx + path_str.len() ..]);
+
+    try!(environment.set_value("PATH", &new_path)
+         .map_err(|_| Error::PermissionDenied));
+
+    // Tell other processes to update their environment
+    unsafe {
+        SendMessageTimeoutA(HWND_BROADCAST,
+                            WM_SETTINGCHANGE,
+                            0 as WPARAM,
+                            "Environment\0".as_ptr() as LPARAM,
+                            SMTO_ABORTIFHUNG,
+                            5000,
+                            ptr::null_mut());
+    }
```

Comment:
Oh I realized this unsafe block is duplicated with the one above, so maybe worth making a helper function? Doesn't seem too important.